File: apps/api/src/rekenraam_api/api/v1/notes.py
from fastapi import APIRouter, Depends, HTTPException, status

from rekenraam_api.api.dependencies import get_ergonomics_service
from rekenraam_api.schemas.ergonomics import MarkdownNoteInput, MarkdownNoteSummary
from rekenraam_api.services.access import AuthorizationError
from rekenraam_api.services.ergonomics import ErgonomicsService

router = APIRouter(prefix="/notes", tags=["notes"])


@router.get("/accounts/{account_id}", response_model=list[MarkdownNoteSummary])
async def list_account_notes(
    account_id: int,
    ergonomics_service: ErgonomicsService = Depends(get_ergonomics_service),
) -> list[MarkdownNoteSummary]:
    try:
        return await ergonomics_service.list_account_notes(account_id)
    except AuthorizationError as error:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(error)) from error
    except ValueError as error:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(error)) from error


@router.post("/accounts/{account_id}", response_model=MarkdownNoteSummary)
async def create_account_note(
    account_id: int,
    input: MarkdownNoteInput,
    ergonomics_service: ErgonomicsService = Depends(get_ergonomics_service),
) -> MarkdownNoteSummary:
    try:
        return await ergonomics_service.create_account_note(account_id, input)
    except AuthorizationError as error:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(error)) from error
    except ValueError as error:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(error)) from error


@router.get("/transactions/{transaction_id}", response_model=list[MarkdownNoteSummary])
async def list_transaction_notes(
    transaction_id: int,
    ergonomics_service: ErgonomicsService = Depends(get_ergonomics_service),
) -> list[MarkdownNoteSummary]:
    try:
        return await ergonomics_service.list_transaction_notes(transaction_id)
    except AuthorizationError as error:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(error)) from error
    except ValueError as error:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(error)) from error


@router.post("/transactions/{transaction_id}", response_model=MarkdownNoteSummary)
async def create_transaction_note(
    transaction_id: int,
    input: MarkdownNoteInput,
    ergonomics_service: ErgonomicsService = Depends(get_ergonomics_service),
) -> MarkdownNoteSummary:
    try:
        return await ergonomics_service.create_transaction_note(transaction_id, input)
    except AuthorizationError as error:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(error)) from error
    except ValueError as error:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(error)) from error


@router.put("/{note_id}", response_model=MarkdownNoteSummary)
async def update_note(
    note_id: int,
    input: MarkdownNoteInput,
    ergonomics_service: ErgonomicsService = Depends(get_ergonomics_service),
) -> MarkdownNoteSummary:
    try:
        note = await ergonomics_service.update_note(note_id, input)
    except AuthorizationError as error:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(error)) from error
    if note is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="note not found")
    return note


@router.delete("/{note_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_note(
    note_id: int,
    ergonomics_service: ErgonomicsService = Depends(get_ergonomics_service),
) -> None:
    try:
        deleted = await ergonomics_service.delete_note(note_id)
    except AuthorizationError as error:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(error)) from error
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="note not found")

File: src/lib/api/notes.ts
import { apiDelete, apiGet, apiPost, apiPut } from "$lib/api/client";

export type MarkdownNote = {
  id: number;
  book_id: number;
  target_type: "account" | "transaction";
  account_id: number | null;
  transaction_id: number | null;
  title: string;
  body_markdown: string;
  pinned: boolean;
  created_at: string;
  updated_at: string;
  created_by_user_id: number | null;
  updated_by_user_id: number | null;
};

export type MarkdownNoteInput = {
  title: string;
  body_markdown: string;
  pinned: boolean;
};

export async function listAccountNotes(accountId: number): Promise<MarkdownNote[]> {
  return apiGet<MarkdownNote[]>(`/notes/accounts/${accountId}`);
}

export async function createAccountNote(accountId: number, input: MarkdownNoteInput): Promise<MarkdownNote> {
  return apiPost<MarkdownNote, MarkdownNoteInput>(`/notes/accounts/${accountId}`, input);
}

export async function listTransactionNotes(transactionId: number): Promise<MarkdownNote[]> {
  return apiGet<MarkdownNote[]>(`/notes/transactions/${transactionId}`);
}

export async function createTransactionNote(transactionId: number, input: MarkdownNoteInput): Promise<MarkdownNote> {
  return apiPost<MarkdownNote, MarkdownNoteInput>(`/notes/transactions/${transactionId}`, input);
}

export async function updateNote(noteId: number, input: MarkdownNoteInput): Promise<MarkdownNote> {
  return apiPut<MarkdownNote, MarkdownNoteInput>(`/notes/${noteId}`, input);
}

export async function deleteNote(noteId: number): Promise<void> {
  await apiDelete<void>(`/notes/${noteId}`);
}
